tests/test_scrollable.py

```diff
@@ -189,5 +190,136 @@ def test_negative(self):
+
+        widget.keypress(reduced_size, "end")
+
+        self.assertEqual(
+            (
+                "License along with this library; if     ",
+                "not, write to the Free Software         ",
+                "Foundation, Inc., 51 Franklin Street,   ",
+                "Fifth Floor, Boston, MA  02110-1301     ",
+                "USA                                    █",
+            ),
+            widget.render(reduced_size).decoded_text,
+        )
+
+        widget.keypress(reduced_size, "home")
+
+        self.assertEqual(top_position_rendered, widget.render(reduced_size).decoded_text)
+
+        widget.mouse_event(reduced_size, "mouse press", 5, 1, 1, False)
+
+        self.assertEqual(pos_1_down_rendered, widget.render(reduced_size).decoded_text)
+
+        widget.mouse_event(reduced_size, "mouse press", 4, 1, 1, False)
+
+        self.assertEqual(top_position_rendered, widget.render(reduced_size).decoded_text)
+
+    def test_large_non_selectable(self):
+        top = urwid.Text("\n".join(string.ascii_letters))
+        bottom = urwid.Text("\n".join(string.digits))
+        widget = urwid.ScrollBar(urwid.ListBox(urwid.SimpleListWalker((top, bottom))))
+
+        reduced_size = (3, 5)
+
+        top_position_rendered = ("a █", "b  ", "c  ", "d  ", "e  ")
+        pos_1_down_rendered = ("b  ", "c █", "d  ", "e  ", "f  ")
+
+        self.assertEqual(top_position_rendered, widget.render(reduced_size).decoded_text)
+
+        widget.keypress(reduced_size, "down")
+
+        self.assertEqual(pos_1_down_rendered, widget.render(reduced_size).decoded_text)
+
+        widget.keypress(reduced_size, "up")
+
+        self.assertEqual(top_position_rendered, widget.render(reduced_size).decoded_text)
+
+        widget.keypress(reduced_size, "page down")
+
+        self.assertEqual(
+            ("f  ", "g █", "h  ", "i  ", "j  "),
+            widget.render(reduced_size).decoded_text,
+        )
+
+        widget.keypress(reduced_size, "page up")
+
+        self.assertEqual(top_position_rendered, widget.render(reduced_size).decoded_text)
+
+        widget.mouse_event(reduced_size, "mouse press", 5, 1, 1, False)
+
+        self.assertEqual(pos_1_down_rendered, widget.render(reduced_size).decoded_text)
+
+        widget.mouse_event(reduced_size, "mouse press", 4, 1, 1, False)
+
+        self.assertEqual(top_position_rendered, widget.render(reduced_size).decoded_text)
+
+    def test_large_selectable(self):
+        """Input is handled by LineBox and wrapped widgets."""
+        top = urwid.Edit("\n".join(string.ascii_letters))
+        bottom = urwid.IntEdit("\n".join(string.digits))
+        widget = urwid.ScrollBar(urwid.ListBox(urwid.SimpleListWalker((top, bottom))))
+
+        reduced_size = (3, 10)
+
+        top_position_rendered = ("a █", "b █", "c  ", "d  ", "e  ", "f  ", "g  ", "h  ", "i  ", "j  ")
+        pos_1_down_rendered = ("0  ", "1  ", "2  ", "3  ", "4  ", "5  ", "6  ", "7  ", "8 █", "9 █")
```

Comment:
keys are handled by `ListBox` contents and if not handled - by `ListBox`.
Smooth scroll was not announced, only display position.